### PR TITLE
add db column for storing user ID, and associated endpoint for looking up referrals by ID

### PR DESCRIPTION
**Pull request by tomsmyers**
## What does this pull request do?

add db column for storing user ID, and associated endpoint for looking up referrals by ID. no changes to existing api endpoints in this PR.

## What is the intent behind these changes?

working towards IC-680

### CODE REVIEW

**Review by karenmillermoj — APPROVED** [review]
Everything looks good

File: src/main/resources/db/migration/V1_1__add_created_by_user_id.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+alter table referral
+    add column created_by_userid text;
```
Comment (sldblog) [inline src/main/resources/db/migration/V1_1__add_created_by_user_id.sql:2]: For the data users we will have to establish where these users come from. Is it possible to capture the auth source here, either as a separate column, table or JSONB payload?

**Review by sldblog — COMMENTED** [review]
LGTM, one question

File: src/main/resources/db/migration/V1_1__add_created_by_user_id.sql (hunk 1 above)
Comment (sldblog) [reply src/main/resources/db/migration/V1_1__add_created_by_user_id.sql:2]: Or by encoding it into the name and switching to integer?

`created_by_delius_user_id integer`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/repository/ReferralRepository.kt [hunk 2]
```diff
@@ -4,4 +4,6 @@ import org.springframework.data.repository.CrudRepository
 import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Referral
 import java.util.UUID
 
-interface ReferralRepository : CrudRepository<Referral, UUID>
+interface ReferralRepository : CrudRepository<Referral, UUID> {
+  fun findByCreatedByUserID(userId: String): List<Referral>
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/repository/ReferralRepository.kt:8]: Can we create an index for this lookup, please? It'll be a major use case to filter the table by user.

**Review by sldblog — COMMENTED** [review]
Oops, forgot an index question

File: src/main/resources/db/migration/V1_1__add_created_by_user_id.sql (hunk 1 above)
Comment (tomsmyers) [reply src/main/resources/db/migration/V1_1__add_created_by_user_id.sql:2]: do we need to store that info, or just output in our events?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/repository/ReferralRepository.kt (hunk 2 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/repository/ReferralRepository.kt:8]: yep, i'll create a ticket and address is a seperate PR

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/repository/ReferralRepository.kt:8]: scratch that, i'll just do it now.

File: src/main/resources/db/migration/V1_1__add_created_by_user_id.sql (hunk 1 above)
Comment (sldblog) [reply src/main/resources/db/migration/V1_1__add_created_by_user_id.sql]: I think so... In this use case:
- Someone looking at our data outside our application for reporting purposes
- They need to establish who this identity is
- We don't have a user table -- so this is an external reference -- but there's no context where it points

**Review by sldblog — APPROVED** [review]
👍 LGTM

I think the user context is worth discussing but isn't a blocker

**Pull request merged**